### ISSUE TITLE
False positive for a Literal with multiple options

### ISSUE THREAD

**Issue by GabrielBianconi** [microsoft/pyright#10405]
The following code causes a complaint about `foo` but not a complaint about `bar` even though they're equivalent:

```py
from typing import Literal, Union

from typing_extensions import TypedDict


X = TypedDict("X", {"value": Literal["x"]})

Y = TypedDict("Y", {"value": Literal["y"]})

XY = Union[X, Y]

Value = Literal["x", "y"]


# This gives me a pyright error
def foo(value: Value) -> XY:
    return {"value": value}


# This is fine...?
def bar(value: Value) -> XY:
    if value == "x":
        return {"value": value}
    else:
        return {"value": value}


def main():
    x = foo("x")
    print(x)

    y = bar("y")
    print(y)


if __name__ == "__main__":
    main()
```

```bash
/main.py:17:12 - error: Type "dict[str, str]" is not assignable to return type "XY"
    Type "dict[str, str]" is not assignable to type "XY"
      "dict[str, str]" is not assignable to "X"
      "dict[str, str]" is not assignable to "Y" (reportReturnType)
1 error, 0 warnings, 0 informations
```

I'm using uv, Python 3.10, PyRight 1.1.400, typing-extensions 4.13.2. Running with `uv run pyright`.

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug. Mypy's behavior is the same.

The type of expression `{"value": value}` must evaluate to a specific type at construction time. A constructor cannot create a union of types. In your example, `bar` works because both `{"value": value}` expressions have an unambiguous type.

**Comment by GabrielBianconi** [comment]
For context, this comes up in our project because the OpenAI client uses something similar to `XY` for different message roles (in simple terms, `Message` is a union of `UserMessage`, `AssistantMessage`, etc.; each has a Literal for a role; and there's a union of those literals elsewhere).

https://github.com/tensorzero/tensorzero/pull/2013/files#diff-47a70acd90bd468d4fbf36b3ef60e5377304564bc647cc02d90b039e8ceef009R204-R214
